JENKINS-37104# Matrix project handling

vivek opened the pull request:
# Description

See https://issues.jenkins-ci.org/browse/JENKINS-37104

# Submitter checklist
- [x] Link to JIRA ticket in description, if appropriate.
- [x] Change is code complete and matches issue description
- [x] Apppropriate unit or acceptance tests or explaination to why this change has no tests
- [x] Reviewer's manual test instructions provided in PR description. See Reviewer's first task below.

- Create a matrix project mp1 with two axis with two values each and run it, you should get total of 4 builds, one for each combination and there should be build for mp1 as well.
- Do curl -v http://localhost:8080/jenkins/blue/rest/search/?q=type:pipeline;excludedFromFlattening:jenkins.branch.MultiBranchProject,hudson.matrix.MatrixProject to get list of pipelines, here we are asking not to inline MatrixProject children.
- The list of pipeline should have mp1 project with self href pointing to "/job/mp1/"
- curl -v http://localhost:8080/jenkins/blue/rest/classes/io.jenkins.blueocean.rest.impl.pipeline.MatrixProjectImpl/ should give list of capabilities that includes hudson.matrix.MatrixProject

- [x] Ran Acceptance Test Harness against PR changes.
https://ci.blueocean.io/job/Acceptance%20Tests%20Param/19/, it failed with unrelated error to download jquery2:

    [ERROR - 2016-08-17T01:17:57.972Z] Session [0f406550-6418-11e6-9fd4-d5cee85dbc1a] - page.onError - msg: Module load failure: Timed out waiting on module 'jquery-detached:jquery2' to load.

Triggered another: https://ci.blueocean.io/job/Acceptance%20Tests%20Param/20/


# Reviewer checklist
- [ ] Run the changes and verified the change matches the issue description
- [x] Reviewed the code
- [x] Verified that the appropriate tests have been written or valid explaination given

@jenkinsci/code-reviewers @reviewbybees 

- Bare minimum support for Matrix project. Ensures it doesn't break.
- For matrix project, self href points to classic URL path ("/job/:name")
- Adds hudson.matrix.MatrixProject capability
- UI to check hudson.matrix.MatrixProject capability and if present create a link using returned self href,
  so that when user clicks on it, he gets redirected to classic Matrix job page.

reviewbybees commented:
This pull request originates from a [CloudBees](https://www.cloudbees.com/) employee. At CloudBees, we require that all pull requests be reviewed by other CloudBees employees before we seek to have the change accepted. If you want to learn more about our process please see [this explanation](https://github.com/reviewbybees/about#about-reviewbybees).

michaelneale commented:
@vivek seems to have upset the favoriting feature (cc @cliffmeyers) - kicking it again to see if sporadic.

michaelneale commented:
🐛  fails the fav. tests with 500 errors, but not sure why. Can you get @cliffmeyers to cast eye over it?

vivek commented:
@michaelneale I removed test dependency on matrix project plugin thinking we don't need it. This broke the tests as Favorite plugin depends on matrix project plugin. I reverted that dependency and tests pass now. Is there anything holding back this merge?

michaelneale commented:
@vivek I think it should be ok - myself or someone else just needs to eyeball it (but I won't get to it this morning). If anyone else can try this one out and check it is ok, would be appreciated.

imeredith commented:
@vivek I cant test out the code right now, but where is the pick that makes the following work?

* UI to check hudson.matrix.MatrixProject capability and if present create a link using returned self href, so that when user clicks on it, he gets redirected to classic Matrix job page.

vivek commented:
@imeredith 'but where is the pick that makes the following work?' you mean mockup?

imeredith commented:
@vivek Sorry i meant 'part'. I dont see code that makes UI redirect if matrix job is clicked?

vivek commented:
@imeredith Oh, its to be done in separate ticket, not part of this PR, https://issues.jenkins-ci.org/browse/JENKINS-37427.

michaelneale commented:
@vivek ready for review (can we remove the work-in-progress label?)

imeredith commented:
Ok 🐝  from me then.

vivek commented:
@imeredith I take that as ready to merge?

michaelneale commented:
I won't have a chance to eyeball this today, so if @imeredith can or @vivek you are happy it loads up blueocean ui with some existing jobs etc - then I am ok with it. Code looks good. ATH passes.

vivek commented:
@michaelneale yes I have done a clean build, loads up fine with different kind of jobs. Going to merge.

basil commented on blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/MatrixProjectImpl.java:
```diff
@@ -0,0 +1,101 @@
+ */
+
+import hudson.Extension;
+import hudson.matrix.MatrixProject;
```
Neglected to add the corresponding dependency to `blueocean-pipeline-api-impl/pom.xml`, implicitly relying on `htmlpublisher` to provide it (no longer the case after https://github.com/jenkinsci/htmlpublisher-plugin/pull/287).
